(PUP-5475) Refactor windows service provider to connect directly to windows API

McdonaldSeanp opened the pull request:
Neither net.exe or sc.exe really satisfy our management needs for windows services, so this PR adds a new Util::Windows::Service ruby module to puppet's utilities that implements service control on windows directly with the windows API via FFI.

Subsequently, this PR updates the service provider to re-implement the various functions of the provider API using the Windows::Service utility rather than win32-service + net.exe

McdonaldSeanp commented:
jenkins please test this on windows2008r2-64a

McdonaldSeanp commented:
This PR is still VERY WIP.

It's up now for PR testing and early discussion

puppetcla commented:
CLA signed by all contributors.

ekinanp commented:
Thanks for putting this up, Sean! I spent some time looking into this ticket so that I have some context on what's going on here. I have my notes pushed up here if they'd be useful to you: https://github.com/ekinanp/puppet/commit/PUP-5475. Some questions/comments:

* Have you looked into the Win32::Service codebase? If so, could we use some of its API calls here? From what I've seen, I don't think so b/c they use the ANSI system calls (E.g. OpenServiceA) while we're using unicode (OpenServiceW). Plus, I kind of like your approach of just creating a separate FFI module. Gives us greater flexibility.

* We don't have any acceptance tests for the Windows service provider. The closest thing I could find is https://github.com/puppetlabs/puppet/blob/master/spec/integration/provider/service/windows_spec.rb and those tests don't really test any functionality. They mostly just test to make sure the results returned by the service provider make sense. I think we should add some as part of this ticket, especially once we work on some of the other service improvements in the epic + any future service improvements. It looks like Windows Powershell lets you create services directly from what I'm seeing in https://msdn.microsoft.com/en-us/magazine/mt703436.aspx so we can use that to create stub services we can delete after the tests. I'm interested in looking into this since it's work that can be done independently of what you're doing here.

* How are you testing the FFI code? Are you testing it interactively as you develop it in e.g. an interpreter?

ekinanp commented on lib/puppet/provider/service/windows.rb:
```diff
@@ -73,7 +73,10 @@ def start
     end
 
-    net(:start, @resource[:name])
+    Puppet::Util::Windows::Service.open_service(@resource[:name]) do |service|
```
I don't think we should let the service provider do this type of low-level manipulation of service handlers. It would probably be better to have public API methods like:
```
Windows::Service.start(service_name)
Windows::Service.stop(service_name)
Windows::Service.status(service_name)
```

that we can call out to from here.

McdonaldSeanp replied:
Yeah, I was trying to allow as much flexibility in the util module, but I think you're right about this.

ekinanp commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,250 @@
+  module Service
+    extend FFI::Library
+
+    # Service type codes
```
Maybe we could split this up? Something like:
```
windows/service/ffi_constants.rb
windows/service/ffi_functions.rb
windows/service/ffi_structs.rb
windows/service.rb -- Contains the public API
```

McdonaldSeanp commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,250 @@
+    def open_service(service_name)
+      scm = OpenSCManagerW(nil, nil, SC_MANAGER_ALL_ACCESS)
+      service = OpenServiceW(scm, service_name, SC_MANAGER_ALL_ACCESS)
+      raise Puppet::Util::Windows::Error.new(_("Failed to open a handle to the service")) if service == 0
```
That's already what `Puppet::Util::Windows::Error` does. It leverages `FFI::errno` which ffi itself makes a call to GetLastError: https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/windows/error.rb#L13

that error class will call out to https://docs.microsoft.com/en-us/windows/desktop/api/winbase/nf-winbase-formatmessage to add the actual message Windows returns from the error code

joshcooper replied:
I think FFI has a null constant instead of using `0`.

McdonaldSeanp commented:
@ekinanp per the Win32::service code: we are trying to remove the Win32 gems entirely from puppet. They are old gems for which some might still be active (I think chef might own the service one now) but for the most part we have removed most of them due to their inactivity. 

IMO we should keep consistent with the other windows providers and just leverage FFI and the windows API directly, that way we aren't splitting between some providers using FFI and others using gems we aren't sure about the state of.

joshcooper commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,250 @@
+    # @yieldparam [:handle] service the windows native handle used to access
+    #   the service
+    def open_service(service_name)
+      scm = OpenSCManagerW(nil, nil, SC_MANAGER_ALL_ACCESS)
```
`OpenSCManagerW` can fail if the caller doesn't have sufficient permissions. Also, we may need to explicitly enable permissions like we do when managing file dacls.

McdonaldSeanp commented:
jenkins please test this on windows2008r2-64a

ekinanp commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,437 @@
+    # Start a windows service, assume that the service is already in the stopped state
+    #
+    # @param [:string] service_name name of the service to start
+    def start(service_name)
```
Should an exception be thrown here in case the service is `PAUSE` or `PAUSE_PENDING` ? `PAUSE` => `RUNNING` are handled via. a `resume` or `continue` method.

McdonaldSeanp replied:
StartServiceW will throw the exception for us, and likely with a better message.

ekinanp commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,437 @@
+    # @param [:handle] service handle to the service to wait on
+    # @return [bool] 'true' once the service is reporting SERVICE_STOPPED,
+    #   'false' if the service was not waiting to stop or queries failed
+    def self.wait_for_service_pending(service, pending_state, complete_state)
```
Maybe rename to `wait_for_pending_transition(service, pending_state, final_state)`? Also, I think this would be better to read if it was written as:
```
loop do
  state = service_state(service)
  return if state == final_state
  unless state == pending_state
    // Raise error since it's an unknown state ?
  end
  // wait-hint stuff here
end
```

ekinanp commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,437 @@
+    def self.open_service(service_name)
+      scm = OpenSCManagerW(nil, nil, SC_MANAGER_ALL_ACCESS)
+      raise Puppet::Util::Windows::Error.new(_("Failed to open a handle to the service control manager")) if scm == FFI::Pointer::NULL_HANDLE
+      service = OpenServiceW(scm, service_name, SC_MANAGER_ALL_ACCESS)
```
I think the reason the acceptance test is failing is b/c you need to cast `service_name` into a wide-string?

ekinanp commented on acceptance/tests/resource/service/windows.rb:
```diff
@@ -0,0 +1,122 @@
+test_name "Windows Service Provider" do
```
It might be useful to add an acceptance test that ensures Puppet fails if the service's pending transition hasn't made any progress since the last wait hint. The C# class can be modified to change the `Sleep` method to something like `StubPendingTransition(waitHint, duration)` and the fixture can take additional parameters like `onStart_waitHint`, `onStart_duration` for each of the `on` methods to get this to work.

McdonaldSeanp commented:
jenkins please test this on windows2008r2-64a

ekinanp commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,726 @@
+          # don't attempt to fail here if the service isn't started because windows error codes
+          # are likely more informative than ours and a failed call to ControlService will produce
+          # those errors
+          wait_for_pending_transition(service, SERVICE_START_PENDING, SERVICE_RUNNING)
```
Is it necessary to wait for the pending transition? The [state transition chart](https://docs.microsoft.com/en-us/windows/desktop/services/service-status-transitions) seems to indicate that you can stop a service from any state so if this errors, that's something with the service (and not on our end). Plus, we'll need to wait for `CONTINUE_PENDING` anyways as this currently stands, so it'd be nice if it could be simplified.

McdonaldSeanp replied:
Yeah, it is, see https://github.com/puppetlabs/puppet/pull/7017#issuecomment-417805844

ekinanp commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,726 @@
+          if ControlService(service, SERVICE_CONTROL_STOP, status) == FFI::WIN32_FALSE
+            raise Puppet::Util::Windows::Error.new(_("Failed to send stop control to service, current state is %{current_state}. Failed with") % { current_state: status[:dwCurrentState].to_s })
+          end
+          unless wait_for_pending_transition(service, SERVICE_STOP_PENDING, SERVICE_STOPPED)
```
`wait_for_pending_transition` already raises an error.  I think just the error message needs to be updated?

McdonaldSeanp replied:
It actually doesn't raise an error, it returns false. If you think it makes more sense to raise an error I could understand that.

See more explanation on https://github.com/puppetlabs/puppet/pull/7017#issuecomment-417805844

ekinanp commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,726 @@
+      end
+      case status[:dwCurrentState]
+      when SERVICE_STOPPED
+        'SERVICE_STOPPED'
```
I think it's better Ruby-style to return symbols instead of strings for enums.

ekinanp replied:
Could also specify a hash. Maybe something like
```
SERVICE_STATES = {
  :SERVICE_STOPPED => <value>,
  ...
}
```

ekinanp commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,726 @@
+        # in this list is dwStartType (the third param). Thus no
+        # generic function was written to make use of all the params
+        # since the API as-is couldn't use them anyway
+        if ChangeServiceConfigW(
```
I think this would be better to read if the result of `ChangeServiceConfigW` was assigned as a variable.

ekinanp commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,726 @@
+    #                         'service_status_process' => SERVICE_STATUS_PROCESS struct
+    #                       }
+    #   }
+    def enum_services
```
Could this be renamed to `services`? Retrieving the services is what it's doing.

ekinanp commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,726 @@
+        service = FFI::Pointer::NULL_HANDLE
+        open_scm(scm_access) do |scm|
+          FFI::MemoryPointer.from_string_to_wide_string(service_name) do |service_name_ptr|
+            service = OpenServiceW(scm, service_name_ptr, service_access)
```
You could avoid this block by just using `wide_string` like here: https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/windows/security.rb#L170

McdonaldSeanp replied:
Oooooo, yeah, I think you're right since it's type is LPCWSTR. For some reason I thought it was not a constant string

ekinanp commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,726 @@
+        FFI::MemoryPointer.new(size_required) do |ssp_ptr|
+          status = SERVICE_STATUS_PROCESS.new(ssp_ptr)
+          FFI::MemoryPointer.new(:lpword) do |bytes_pointer|
+            if QueryServiceStatusEx(
```
This could also go into the previous `bytes_pointer` block.

ekinanp commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,726 @@
+        end
+        FFI::MemoryPointer.new(size_required) do |ssp_ptr|
+          config = QUERY_SERVICE_CONFIGW.new(ssp_ptr)
+          FFI::MemoryPointer.new(:lpword) do |bytes_pointer|
```
Could also move to the previous `bytes_pointer` block

ekinanp commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,726 @@
+      #
+      # @param [Integer] wait_hint the wait hint of a service in milliseconds
+      # @return [Integer] the time to wait in seconds between querying the service
+      def process_wait_hint(wait_hint)
```
Might be better to rename this `wait_hint_to_wait_time`. `process_wait_hint` makes it seem like there's way more stuff going on in this method.

ekinanp commented on spec/unit/provider/service/windows_spec.rb:
```diff
@@ -151,33 +120,25 @@
-
     # We need to guard this section explicitly since rspec will always
     # construct all examples, even if it isn't going to run them.
     if Puppet.features.microsoft_windows?
```
I don't think this check is necessary anymore.

McdonaldSeanp replied:
I don't understand why, but without this guard rspec will try to run these examples outside windows because they are defined in a loop.

(I actually had to add this guard to some new tests too)

McdonaldSeanp commented:
@ekinanp so, I think `wait_for_pending_transition` begs a little explination:

The way I designed the usage of that function is the following, (using start as an example, stop is just the opposite anyway):

1. wait for the service to stop if and only if it's in STOP_PENDING. If the service _is_ in STOP_PENDING, it won't accept any control codes sent to it (including the call to StartService). If the service _is not_ in stop pending or stopped (i.e. if it's paused), just continue and let StartService error with windows messages.
2. execute StartService (or ControlService for the stop command), if the command fails use the Windows::Error util to report GetLastError()
3. wait for the service to move from START_PENDING to RUNNING, if the service isn't in those states then actually fail this time that the service wasn't really started

The reason we need to wait the first time on the STOP_PENDING (or START_PENDING for stop()) is MS themselves suggest that services _do not_ accept any control codes or controls in pending states. You can see in the 'best practices' list under https://docs.microsoft.com/en-us/windows/desktop/api/Winsvc/nf-winsvc-setservicestatus#remarks that:

"Do not register to accept controls while the status is SERVICE_START_PENDING or the service can crash. After initialization is completed, accept the SERVICE_CONTROL_STOP code."

and I assume the same goes for SERVICE_STOP_PENDING.

Even in the provided examples from MS like: https://docs.microsoft.com/en-us/windows/desktop/Services/starting-a-service https://docs.microsoft.com/en-us/windows/desktop/Services/stopping-a-service they suggest waiting on the pending transitions of the opposite call (so wait on STOP_PENDING for start(), etc.).

Beyond that reasoning, I figured it would be easier to let StartService and ControlService error when the service was in anything other than stop/start and stop-pending/start-pending rather than fail on the first calls to `wait_for_pending_transition`, which is why that function returns a boolean false when the state doesn't match pending_state or final_state: https://github.com/puppetlabs/puppet/pull/7017/files#diff-cbf3560d761defff8b026335352495abR553

That might make sense to change, I can always just have it error there instead of returning false if you think that makes more sense.

McdonaldSeanp commented:
jenkins please test this on windows2008r2-64a

ekinanp commented:
@McdonaldSeanp Ohhh I missed the boolean returns -- brain fart on my part. Ok, that makes sense. Thank you for clarifying this. I can understand why too -- `START_PENDING` is when the thread's being created while `STOP_PENDING` is when it is being killed. `PAUSE_PENDING` + `CONTINUE_PENDING` is when the thread's already created so those can safely accept the `STOP` signal.

I do think that in `stop`, we should probably not raise an error if the service is in any other state but `START_PENDING` (I'm ignoring `STOP_PENDING` and `STOPPED` b/c Puppet will not call `stop` if the service is already in those states). I think this is the correct approach based on the transition diagram, and the fact that the docs you linked don't mention there being any issues with sending control signals in those other states. Also, there is a Windows ticket about managing paused services -- https://tickets.puppetlabs.com/browse/PUP-6702. My gut feeling tells me the work for that might require a new `ensure` value (`paused`) so it might be good to handle it here while the refactor's still ongoing.

McdonaldSeanp commented:
@ekinanp Per the paused states and ticket: I would like to leave work for paused states out of this PR, and leave that ticket as work for someone once this work is merged. That way I can keep this PR as focused on one thing (the migration to FFI) as possible

I think we will need to have a discussion about if it's appropriate to land the paused states work in 5.5.x anyway, so I was going to leave that until after this was done.

McdonaldSeanp commented:
jenkins please test this on windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2012r2_ja-64a%7Blocale=ja%7D,windows2016-64a

McdonaldSeanp commented:
jenkins please test this on windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2016-64a

McdonaldSeanp commented:
The development part of this PR is ready to go, the only thing left is to finalize what to do with Japanese windows testing.

McdonaldSeanp commented:
I am making my way through acceptance/unit testing for services with localized names now, so this remains blocked.

However the functional code should be more or less complete.

speedofdark commented on spec/unit/util/windows/service_spec.rb:
```diff
@@ -0,0 +1,674 @@
+
+  # the following is a helper that adds both a call to SERVICE_STATUS_PROCESS.new
+  # and QueryServiceStatusEx. What this does from the perspective of the tests is
+  # emulate would would happen from a succesful call to QueryServiceStatusEx that
```
nit: emulate would would happen -- multiple instances :) ...

McdonaldSeanp commented:
testing is finished, this is ready for merge.

McdonaldSeanp commented:
jenkins please test this on windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2012r2_ja-64a%7Blocale=ja%7D,windows2016-64a

McdonaldSeanp commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,708 @@
+          # those errors
+          wait_for_pending_transition(service, SERVICE_START_PENDING, SERVICE_RUNNING)
+          if ControlService(service, SERVICE_CONTROL_STOP, status) == FFI::WIN32_FALSE
+            raise Puppet::Util::Windows::Error.new(_("Failed to send stop control to service, current state is %{current_state}. Failed with") % { current_state: status[:dwCurrentState].to_s })
```
`Puppet::Util::Windows::Error` will append the error message I gave it with the actual error message reported by getlasterror: https://msdn.microsoft.com/en-us/library/windows/desktop/ms679360(v=vs.85).aspx

So the error message will look something like:

```
Failed to send stop control to service, current state is RUNNING. Failed with : windows error message
```

speedofdark replied:
Sweet.

McdonaldSeanp commented on acceptance/tests/resource/service/windows_mixed_utf8.rb:
```diff
@@ -0,0 +1,88 @@
+        assert_service_properties_on(agent, mock_service[:name], State: 'Running')
+      end
+
+      step 'Verify idempotence' do
```
Yeah it should be a noop, the check for idempotence (at least in my mind) is that the second call to the start operation will call on the same service that should still be in the running state (and thus create a noop) from the first call.

To me this is checking that we are actually calling on a working service that should behave idempotently when the same status `ensure => :running` is set multiple times.

speedofdark replied:
Makes sense.

McdonaldSeanp commented on acceptance/tests/resource/service/windows_mixed_utf8.rb:
```diff
@@ -0,0 +1,88 @@
+test_name "Windows Service Provider With Mixed UTF-8 Service Names" do
+  confine :to, :platform => 'windows'
```
Yeah, I was sort of thinking that too. Now that someone else had the same thought I think I'll remove the JP iteration and just leave the UTF-8 mixed one.

Iristyle commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,708 @@
+
+    # Service control codes
+    # https://docs.microsoft.com/en-us/windows/desktop/api/Winsvc/nf-winsvc-controlservice
+    SERVICE_CONTROL_CONTINUE       = 0x00000003
```
I'm sure everything here for services is probably fine, but you might want to double check the constants against the latest copy of the Windows 10 SDK. We found values missing from MSDN, but that were present in the actual headers during the scheduled task rewrite.

In particular some of the latter version constants in https://github.com/puppetlabs/puppetlabs-scheduled_task/blob/master/lib/puppet_x/puppetlabs/scheduled_task/task.rb#L29-L41 were simply missing from online docs.

Iristyle commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,708 @@
+    # } ENUM_SERVICE_STATUS_PROCESSW, *LPENUM_SERVICE_STATUS_PROCESSW;
+    class ENUM_SERVICE_STATUS_PROCESSW < FFI::Struct
+      layout(
+        :lpServiceName, :pointer,
```
We have an alias for `:lpwstr` already that you should use - see https://github.com/puppetlabs/puppet/blob/ba4d1a1aba0095d3c70b98fea5c67434a4876a61/lib/puppet/util/windows/api_types.rb#L139

Ditto for all other instances

McdonaldSeanp replied:
I tried that and the FFI:Struct class failed everywhere

McdonaldSeanp replied:
```
Failure/Error:
  layout(
    :lpServiceName, :lpwstr,
    :lpDisplayName, :pointer,
    :ServiceStatusProcess, SERVICE_STATUS_PROCESS
  )

TypeError:
  invalid struct field type #<FFI::Type::Builtin:BUFFER_INOUT size=8 alignment=8>
```

Iristyle replied:
That's interesting. Wonder if it's an FFI bug? Or desired behavior specific to `Struct`?

From an archived post at https://ffi.github.io/ruby-ffi-archive/messages/20130310-Re_%5Bruby-ffi%5D%20Passing%20a%20struct%20by%20value_%20and%20typedef-1080.html :

```
:buffer_in, and :buffer_out are optimization hints for VMs when they need to copy data from VM heap memory to native memory:

:buffer_in means "allocate some temporary native memory, copy the contents of this object to it, then pass the temporary memory address to the function"

:buffer_out means "allocate some temporary native memory, pass that address to the function, then afterwards, copy the contents of the temporary memory area back to the VM heap object"

:pointer and :buffer_inout are analogous and mean "do both of the above" - i.e. copies both in to and out of the temporary memory area.
```

Iristyle commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,708 @@
+        :dwServiceType, :dword,
+        :dwStartType, :dword,
+        :dwErrorControl, :dword,
+        :lpBinaryPathName, :pointer,
```
`:lpwstr`

Iristyle commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,708 @@
+      # @param [Integer] final_state the state indicating the transition is finished
+      # @return [bool] 'true' once the service is reporting final_state,
+      #   'false' if the service was not in pending_state or finaL_state
+      def wait_for_pending_transition(service, pending_state, final_state)
```
Can / should this be user configurable, rather than relying on `:dwWaitHint`?

McdonaldSeanp replied:
That is a good question. I was actually thinking about this, and I'm not sure which path to take with this scenario for puppet 5.5.x

I think in puppet 6 we should probably make it user configurable (at least optionally), but I'm unsure if that is too dangerous a change for a Z release (and thus we should keep it out of 5.5.x)

Since we were using net.exe before there was no wait on pending anyways, and using the wait hint is better (IMO) than just trying to guess what a good time is.  This is the recommended way MS describes in their examples on how to wait for a pending transition.

But the question does remain if we think it would be safe enough to land what is essentially a small new feature in a Z release or if it might effect the stability of the product.

What do you think @e?

Iristyle replied:
I agree with you that it's safe as-is. Maybe file a ticket on this one.

I just wasn't sure (not having read the docs yet) if it's possible to get stuck here infinitely waiting on an unresponsive service.

McdonaldSeanp replied:
That shouldn't be possible, I can look in to a test for this but the logic is built in such a way that unless the service reports an increment to dwCheckPoint puppet fails after the wait hint elapses, so unless someone builds their service logic to infinitely increment the checkpoint it should always fail at some point.

I might file two tickets here: 1. add a new feature to puppet 6 to allow for configurable timeout, and 2. add a base timeout that stops the logic after some long period no matter what ( I think, for example, the scm will bail on a service after something like two minutes regardless of wait hint or check point)

McdonaldSeanp replied:
Ah, I forgot about that part: yeah MS themselves in their examples suggest this: the master minimum wait time is at least 1 second, and the master maximum wait time is 10 seconds. Here is the example I'm referring to: https://docs.microsoft.com/en-us/windows/desktop/Services/starting-a-service

Iristyle commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,708 @@
+    #  windows service, see the "Service start type codes" section in the
+    #  Puppet::Util::Windows::Service file for the list of available codes
+    def set_startup_mode(service_name, startup_type)
+      startup_code = SERVICE_START_TYPES.invert[startup_type]
```
I think you want `.key` here, which doesn't create Ruby object garbage by inverting / creating a new hash every time and performing the lookup against that.

Iristyle commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,708 @@
+        # are likely more informative than ours and a failed call to StartServiceW will produce
+        # those errors
+        wait_for_pending_transition(service, SERVICE_STOP_PENDING, SERVICE_STOPPED)
+        if StartServiceW(service, 0, nil) == FFI::WIN32_FALSE
```
Rather than Ruby `nil`, as a matter of style I usually use `FFI::Pointer::NULL`. It can be another clue to readers of the code that its a Win32 API call.

Iristyle commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,708 @@
+      ensure
+        CloseServiceHandle(service)
+      end
+      private :open_service
```
I'm not sure that `private` actually works in this context.

Iristyle commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,708 @@
+    attach_function_private :EnumServicesStatusExW,
+      [
+        :handle,
+        :int,
```
You can defined enums like this https://github.com/puppetlabs/puppet/blob/ba4d1a1aba0095d3c70b98fea5c67434a4876a61/lib/puppet/util/windows/principal.rb#L149-L161

Iristyle commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,708 @@
+    # );
+    ffi_lib :advapi32
+    attach_function_private :ControlService,
+      [:handle, :dword, :lpvoid], :win32_bool
```
I think we typically use `:pointer` for pointers to structs.

See https://github.com/puppetlabs/puppet/blob/ba4d1a1aba0095d3c70b98fea5c67434a4876a61/lib/puppet/util/windows/process.rb#L465-L491 for example

Iristyle commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,708 @@
+    # );
+    ffi_lib :advapi32
+    attach_function_private :StartServiceW,
+      [:handle, :dword, :lpwstr], :win32_bool
```
Pointer pointers aren't that common... but we typically define them as `:pointer` like https://github.com/puppetlabs/puppet/blob/ba4d1a1aba0095d3c70b98fea5c67434a4876a61/lib/puppet/util/windows/eventlog.rb#L191

Iristyle commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,708 @@
+    # );
+    ffi_lib :advapi32
+    attach_function_private :QueryServiceConfigW,
+      [:handle, :lpbyte, :dword, :lpword], :win32_bool
```
Even though they both alias to `:pointer, you want `:lpdword` instead of `:lpword`

On the struct pointer we typically use `:pointer`

Iristyle commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,708 @@
+    # );
+    ffi_lib :advapi32
+    attach_function_private :QueryServiceStatusEx,
+      [:handle, :int, :lpbyte, :dword, :lpword], :win32_bool
```
Even though they both alias to `:pointer, you want `:lpdword` instead of `:lpword`

You should be able to use the enum for `SC_STATUS_TYPE`

Iristyle commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,708 @@
+          SERVICE_NO_CHANGE,
+          startup_code,
+          SERVICE_NO_CHANGE,
+          nil,
```
`FFI::Pointer::NULL`

Iristyle replied:
It might be helpful to put comments next to the params like

```ruby
SERVICE_NO_CHANGE, #dwServiceType
startup_code, #dwStartType
SERVICE_NO_CHANGE, #dwErrorControl

... etc
```

I thought this looked suspicious given the same constants were passed to multiple params and had to go back to to the docs to get it...

Iristyle commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,708 @@
+              )
+              size_required = bytes_pointer.read(:dword)
+              FFI::MemoryPointer.new(size_required) do |buffer_ptr|
+                resume_ptr.write_int(0)
```
In practical terms since this value is 0, it doesn't matter... but we do have a `write_dword` alias that you might want to use for style - https://github.com/puppetlabs/puppet/blob/ba4d1a1aba0095d3c70b98fea5c67434a4876a61/lib/puppet/util/windows/api_types.rb#L118

Iristyle commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,708 @@
+                    :display_name => service[:lpDisplayName].read_arbitrary_wide_string_up_to(SERVICENAME_MAX),
+                    :service_status_process => service[:ServiceStatusProcess]
+                  }
+                  cursor_ptr += ENUM_SERVICE_STATUS_PROCESSW.size
```
There's a trick you can use to avoid manually incrementing the pointer like this. See https://github.com/puppetlabs/puppet/blob/ba4d1a1aba0095d3c70b98fea5c67434a4876a61/lib/puppet/util/windows/process.rb#L136-L142

Just note that the indexing is 0-based instead of 1-based like you have here

Iristyle commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,708 @@
+      # @param [Integer] service_access code corresponding to the access type requested for the service
+      # @yieldparam [:handle] service the windows native handle used to access
+      #   the service
+      def open_service(service_name, scm_access, service_access)
```
I think you want the `&block` at the end of the param list to make the block mandatory.

Iristyle commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,708 @@
+      # Opens a handle to the service control manager
+      #
+      # @param [Integer] scm_access code corresponding to the access type requested for the scm
+      def open_scm(scm_access)
```
I think you want the `&block` at the end of the param list to make the block mandatory.

Iristyle commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,708 @@
+        open_scm(scm_access) do |scm|
+          service = OpenServiceW(scm, wide_string(service_name), service_access)
+          raise Puppet::Util::Windows::Error.new(_("Failed to open a handle to the service")) if service == FFI::Pointer::NULL_HANDLE
+          yield(service)
```
nit - `yield` is a keyword, not a method

Iristyle commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,708 @@
+            0,
+            bytes_pointer
+          )
+          size_required = bytes_pointer.read(:dword)
```
Can use `read_dword` here

Iristyle commented on spec/unit/util/windows/service_spec.rb:
```diff
@@ -0,0 +1,667 @@
+  # struct (those that would be set after the call to QueryServiceStatusEx)
+  # when the struct is first created, and the calling process won't be the wiser
+  def expect_successful_status_query_and_return(query_return)
+    FFI::MemoryPointer.expects(:new).with(:lpword).yields(bytes_pointer)
```
Oh wow, you mocked this stuff all out!

Usually for FFI calls we just put stuff in `integration` and makes some real calls against the OS because there's a lot to mock and there's a lot that can go wrong with bad mocks.

McdonaldSeanp replied:
Yeah, I was hoping to move as much out of integration as I could.

McdonaldSeanp replied:
Yeah, I think you might be correct that some of the tests are working through implementation rather than behavior. I am attempting to move things to unit testing as much as possible because unit testing is cheaper.

Iristyle replied:
I don't think we should hold up the PR anymore at this point, but I'm with @glennsarti on integration vs unit. The problem with all the mocks it that they're fragile, rely on internal implementation details and are typically difficult to maintain long term.

When you have logic that you want to test independently, it can often be useful to break out additional pure Ruby classes, as those are easy to unit test quickly if cleanly defined (and without implicit dependencies, which because of Rubys dynamic nature can sometimes make difficult to find).

An example of where we took this tactic recently is in the scheduled task rewrite. The original code had a single class that combined calling the COM APIs, manipulating / validating triggers (https://github.com/puppetlabs/puppet/blob/5.5.x/lib/puppet/util/windows/taskscheduler.rb#L941-L1045), and a bunch of other stuff that was very tightly coupled. It was very unwieldy to rationalize or test as a result.

In addition to a pretty light type and provider, we ended up with a couple of new classes:

* [Trigger](https://github.com/puppetlabs/puppetlabs-scheduled_task/blob/master/lib/puppet_x/puppetlabs/scheduled_task/trigger.rb) has a number of new classes that handle various types of triggers and conversions. It's easy to test in pure Ruby like https://github.com/puppetlabs/puppetlabs-scheduled_task/blob/master/spec/unit/puppet_x/puppetlabs/scheduled_task/trigger_spec.rb
* [Task](https://github.com/puppetlabs/puppetlabs-scheduled_task/blob/master/lib/puppet_x/puppetlabs/scheduled_task/task.rb) does the COM integration and handles the operations on tasks themselves (and setting some of the metadata). We were able to do a lighter integration testing layer in https://github.com/puppetlabs/puppetlabs-scheduled_task/blob/master/spec/integration/puppet_x/puppetlabs/scheduled_task/task_spec.rb because most of the code that needed to be tested was in round-tripping trigger definitions through manifests, etc. (We still have some work to do in refactoring some *other* tests)

Just some food for thought on choosing abstractions that lead to simplified testing. Usually when you have to mock a ton to validate something, it can be an opportunity to rethink these bits to avoid that.

Iristyle commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,708 @@
+    # @param [Int] startup_type a code corresponding to a start type for
+    #  windows service, see the "Service start type codes" section in the
+    #  Puppet::Util::Windows::Service file for the list of available codes
+    def set_startup_mode(service_name, startup_type)
```
Yeah, that seems like a good idea to me

McdonaldSeanp commented:
@Iristyle @glennsarti I cross-checked the constants against the Windows 10 SDK, and updated them. I only included the constants used with the functions provided in this PR, so there might be more constants available to other functions but I figured I would leave them out for now to keep this PR focused.

Glenn, I updated the unit tests to be more focused on method behavior rather than testing for correct API invocations, hopefully the tests look a little cleaner.

The private functions are all included inside `class << self` so I _believe_ the private functions should work., just tell me if it doesn't

Ethan, I'm not sure what we should do about :lpwstr causing failures with FFI::Struct. Should we open an issue with FFI?

McdonaldSeanp commented:
jenkins please test this on windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2012r2_ja-64a%7Blocale=ja%7D,windows2016-64a

McdonaldSeanp commented on acceptance/fixtures/MockService.cs:
```diff
@@ -0,0 +1,80 @@
+
+  protected override void OnStart(string [] args) {
+    RequestAdditionalTime(10000);
+    Sleep(8);
```
Oh, yeah, the next commits remove the default sleep(8) for a configured sleep time, and it's set to 0 when the service should be mocked to work correctly, and set to 20 (it won't actually get there) when the service should be mocked to have slow startups/shutdowns

Iristyle replied:
Ah thanks, missed that!

McdonaldSeanp replied:
I will update RequestAdditionalTime to only request a wait time of 2 seconds for slow startup/shutdown

Iristyle commented on lib/puppet/util/windows/service.rb:
```diff
@@ -0,0 +1,754 @@
+    #   LPCWSTR      pszGroupName
+    # );
+    SC_ENUM_TYPE = enum(
+      :SC_ENUM_PROCESS_INFO, 0,
```
Have to love enums with a single value!

McdonaldSeanp commented:
Manual testing of `puppet resource` with mutation is working, this should be GTG

ekinanp commented on acceptance/fixtures/MockService.cs:
```diff
@@ -0,0 +1,80 @@
+    CanPauseAndContinue = true;
+  }
+
+  private void Sleep(int seconds) {
```
Nitpick: Would be nice to change this to `StubPendingTransition` and move the `RequestAdditionalTime` code here.

McdonaldSeanp replied:
ah, sure
